packages/@magic-ext/farcaster/src/index.ts

```diff
@@ -0,0 +1,132 @@
+import { Extension, FarcasterLoginEventEmit, JsonRpcRequestPayload } from '@magic-sdk/commons';
+import type {
+  CreateChannelAPIResponse,
+  AuthenticateAPIResponse,
+  WatchStatusResponse,
+  AuthClientError,
+} from '@farcaster/auth-client';
+import { createAppClient, viemConnector } from '@farcaster/auth-client';
+import { FarcasterPayloadMethod } from './types';
+import { isMobile } from './utils';
+
+type Handle = {
+  on: Handler;
+};
+
+const EVENT = {
+  CHANNEL: 'channel',
+  DONE: 'done',
+  ERROR: 'error',
+} as const;
+
+type EventName = typeof EVENT[keyof typeof EVENT];
+
+interface EventMap {
+  channel: CreateChannelAPIResponse;
+  done: AuthenticateAPIResponse;
+  error: AuthClientError;
+}
+
+interface Handler {
+  (event: 'channel', callback: (params: CreateChannelAPIResponse) => void): Handle;
+  (event: 'done', callback: (params: AuthenticateAPIResponse) => void): Handle;
+  (event: 'error', callback: (params: AuthClientError) => void): Handle;
+}
+
+const DEFAULT_RELAY_URL = 'https://relay.farcaster.xyz';
+const DEFAULT_SIWE_URI = 'https://example.com/login';
+const DEFAULT_TIMEOUT = 60000;
+const DEFAULT_INTERVAL = 500;
+
+export class FarcasterExtension extends Extension.Internal<'farcaster', any> {
+  name = 'farcaster' as const;
+  config: any = {};
+
+  public login = ({ showUI }: { showUI: boolean }): Handle => {
+    const appClient = createAppClient({
+      relay: DEFAULT_RELAY_URL,
+      ethereum: viemConnector(),
+    });
+
+    const channelPromise = appClient.createChannel({
+      siweUri: DEFAULT_SIWE_URI,
+      domain: window.location.host,
+    });
+
+    const statusPromise: WatchStatusResponse = channelPromise.then(({ data }) => {
+      return appClient
+        .watchStatus({
+          channelToken: data.channelToken,
+          timeout: DEFAULT_TIMEOUT,
+          interval: DEFAULT_INTERVAL,
+        })
+        .then((r) => {
+          if (r.isError) {
+            throw r.error;
+          } else {
+            return r;
+          }
+        });
+    });
+
+    let requestPayload: JsonRpcRequestPayload;
+    let channel_token: string;
+
+    const handle: Handle = {
+      on: (event, callback) => {
+        if (event === EVENT.CHANNEL) {
+          (async () => {
+            const { data } = await channelPromise;
+
+            callback(data as any);
+            channel_token = data.channelToken;
+
+            if (isMobile()) {
```

Comment:
I noticed a potential issue when the user is on a mobile device. The `showUI` condition won't have any effect in that case. If that's by design, I'd suggest putting a console warning or info to explain this. Otherwise, people might ask why the flag is not working.

Reply:
yeah it's by design, we don't display QR on mobile device, and just open channel URL.
I will add console info